// mw-automagic-client/lib/automagic.js
'use strict';

const request = require('request');

const log_error = require('debug')('mw-automagic-client:automagic:error');
const log_debug = require('debug')('mw-automagic-client:automagic:debug');


const UPDATE_INFO_URL_J = 'http://www.sony.co.jp/dpt-rp1/check-for-update';
const UPDATE_INFO_URL_U = 'http://www.sony.net/dpt-rp1/check-for-update';
const UPDATE_INFO_URL_C = 'http://www.sony.net/cn/dpt-rp1/check-for-update';

const CATEGORY_ID = 'DR001';

const SERVICE_ID_DEVICE = 'DPTRP1';
const SERVICE_ID_APP_WIN = 'DigitalPaperApp_W';
const SERVICE_ID_APP_MAC = 'DigitalPaperApp_M';

const LANGUAGE_CODE_JA = 'JAPANESE';
const LANGUAGE_CODE_EN = 'ENGLISH';
const LANGUAGE_CODE_ZH = 'CHINESE(SIMPLIFIED)';

const LOCALE_CODE_JP = 'JAPAN';
const LOCALE_CODE_US = 'US';
const LOCALE_CODE_CN = 'CHINA';


class AutoMagic {

  /**
   * デバイスの「Category ID」を取得する。
   */
  static get CATEGORY_ID() {
    return CATEGORY_ID;
  }

  /**
   * デバイスの「Service ID」を取得する。
   */
  static get SERVICE_ID_DEVICE() {
    return SERVICE_ID_DEVICE;
  }

  /**
   * アプリの「Service ID」を取得する。
   */
  static get SERVICE_ID_APP() {
    switch (process.platform) {
      case 'win32':
        return SERVICE_ID_APP_WIN;
      case 'darwin':
        return SERVICE_ID_APP_MAC;
      default:
        const error = new Error('[ERROR] unsupported platform : ' + process.platform);
        log_error(error);
        throw error;
    }
  }

  /**
   * 更新情報 URL を取得します。
   */
  static getUpdateInfoUrl(localeCode, serviceId) {
    let updateInfoUrl = UPDATE_INFO_URL_U;
    if (localeCode === LOCALE_CODE_JP) {
      updateInfoUrl = UPDATE_INFO_URL_J;
    } else if (localeCode === LOCALE_CODE_CN) {
      updateInfoUrl = UPDATE_INFO_URL_C;
    }

    if (serviceId === SERVICE_ID_APP_WIN) {
      updateInfoUrl += '-win';
    } else if (serviceId === SERVICE_ID_APP_MAC) {
      updateInfoUrl += '-mac';
    }

    return updateInfoUrl;
  }

  /**
   * アプリの「アップデート情報ファイル」のURLを取得する。
   */
  static getUrl_AppUpdateInfoXml(localeCode) {
    let updateInfoUrl = this.getUpdateInfoUrl(localeCode, this.SERVICE_ID_APP);

    return updateInfoUrl;
  }

  /**
   * デバイスの「アップデート情報ファイル」のURLを取得する。
   */
  static getUrl_DeviceUpdateInfoXml(skuCode) {
    let localeCode = this.getLocaleCodeBySku(skuCode);
    let updateInfoUrl = this.getUpdateInfoUrl(localeCode, this.SERVICE_ID_DEVICE);

    return updateInfoUrl;
  }

  /**
   * アプリのアップデート確認用の「User-Agent」を取得する。
   */
  static getUserAgentForAppUpdate(appVersion, newVersion) {
    return this.getUserAgent(appVersion, 'DigitalPaperApp', newVersion);
  }

  /**
   * デバイスのアップデート確認用の「User-Agent」を取得する。
   */
  static getUserAgentForDeviceUpdate(appVersion, newVersion) {
    return this.getUserAgent(appVersion, 'DigitalPaper', newVersion);
  }

  /**
   * 「User-Agent」を取得する。
   * @private
   */
  static getUserAgent(appVersion, software, newVersion) {
    if (!newVersion) {
      newVersion = '';
    }
    const debug = process.env.MW_DEBUG_AUTO_MAGIC ? '-DaM_uAS-' : '';
    return 'DPA_' + appVersion + '#' + software + '#' + newVersion + '#00#' + debug + '#http.status.0#';
  }

  /**
   * アプリの言語設定から
   * 「Auto Magic 言語コード」（大文字）を取得する。
   */
  static getLanguageCode(appLanguage) {
    var f = function () {
      switch (appLanguage) {
        case 'Japanese':
          return LANGUAGE_CODE_JA;
        case 'English':
          return LANGUAGE_CODE_EN;
        case 'Chinese(Simplified)':
          return LANGUAGE_CODE_ZH;
        default:
          return null;
      }
    };

    var result = f();
    log_debug('getLanguageCode() => ' + result);

    return result;
  }

  /**
   * Win/Macのロケール情報から
   * 「Auto Magic ロケールコード」（大文字）を取得する。
   */
  static getLocaleCode(osLocale) {
    var f = function () {
      if (osLocale.endsWith('_JP')) {
        return LOCALE_CODE_JP;
      }
      if (osLocale.endsWith('_US')) {
        return LOCALE_CODE_US;
      }
      if (osLocale.endsWith('_CN')) {
        return LOCALE_CODE_CN;
      }
      return null;
    };

    var result = f();
    log_debug('getLocaleCode() => ' + result);

    return result;
  }

  /**
   * デバイスのSKUコードから
   * 「Auto Magic ロケールコード」（大文字）を取得する。
   */
  static getLocaleCodeBySku(skuCode) {
    var f = function () {
      switch (skuCode) {
        case 'J':
          return LOCALE_CODE_JP;
        case 'U':
          return LOCALE_CODE_US;
        case 'C':
          return LOCALE_CODE_CN;
        default:
          return null;
      }
    };

    var result = f();
    log_debug('getLocaleCodeBySku() => ' + result);

    return result;
  }
}


module.exports = AutoMagic;
