// bt-pairing-mw/index.js
/* ******************* */
/* */
/* Copyright 2017 Sony Corporation */
/* */
/* ******************* */

var EventEmitter = require('events').EventEmitter;
var addon = require('./node-bt-pairing-mw.node');

addon.SBluetoothPairing.prototype.__proto__ = EventEmitter.prototype;

/**
 * Enum for Bluetooth device class.
 * @readonly
 * @enum {number}
 */
addon.SBluetoothPairing.prototype.S_BT_PAIRING_DEVICE_CLASS = {
    /** class mask **/
    S_BT_PAIRING_DEVICE_CLASS_MASK_SERVICE: 0x00ffe000,
    S_BT_PAIRING_DEVICE_CLASS_MASK_DEVICE: 0x00001ffc,
    S_BT_PAIRING_DEVICE_CLASS_MASK_MAJOR: 0x00001f00,
    S_BT_PAIRING_DEVICE_CLASS_MASK_MINOR: 0x000000fc,

    /** major service class **/
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_LIMITED_DISCOVERABLE_MODE : 0x00002000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_RESERVED_14 : 0x00004000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_RESERVED_15 : 0x00008000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_POSITIONING : 0x00010000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_NETWORKING  : 0x00020000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_RENDERING   : 0x00040000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_CAPTURING   : 0x00080000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_OBJECT_TRANSFER : 0x00100000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_AUDIO       : 0x00200000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_TELEPHONY   : 0x00400000,
    S_BT_PAIRING_DEVICE_CLASS_SERVICE_INFORMATION : 0x00800000,

    /** major device class **/
    S_BT_PAIRING_DEVICE_CLASS_AUDIO_VIDEO: 0x00000400,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER: 0x00000100,
    S_BT_PAIRING_DEVICE_CLASS_HEALTH: 0x00000900,
    S_BT_PAIRING_DEVICE_CLASS_IMAGING: 0x00000600,
    S_BT_PAIRING_DEVICE_CLASS_MISC: 0x00000000,
    S_BT_PAIRING_DEVICE_CLASS_NETWORKING: 0x00000300,
    S_BT_PAIRING_DEVICE_CLASS_PERIPHERAL: 0x00000500,
    S_BT_PAIRING_DEVICE_CLASS_PHONE: 0x00000200,
    S_BT_PAIRING_DEVICE_CLASS_TOY: 0x00000800,
    S_BT_PAIRING_DEVICE_CLASS_UNCATEGORIZED: 0x00001f00,
    S_BT_PAIRING_DEVICE_CLASS_WEARABLE: 0x00000700,

    /** full device class **/
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_DESKTOP: 0x00000104,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_HANDHELD_PC_PDA: 0x00000110,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_LAPTOP: 0x0000010c,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_PALM_SIZE_PC_PDA: 0x00000114,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_SERVER: 0x00000108,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_UNCATEGORIZED: 0x00000100,
    S_BT_PAIRING_DEVICE_CLASS_COMPUTER_WEARABLE: 0x00000118,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_CELLULAR: 0x00000204,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_CORDLESS: 0x00000208,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_ISDN: 0x00000214,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_MODEM_OR_GATEWAY: 0x00000210,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_SMART: 0x0000020c,
    S_BT_PAIRING_DEVICE_CLASS_PHONE_UNCATEGORIZED: 0x00000200,
};

module.exports = addon;
