// mw-usb-initiator/lib/dpt-serial-port-controller.js
/**
 * Win/MacにUSB接続されたDPTの仮想シリアルポート(CDC-ACM)の一覧を取り扱う。
 * DPTの仮想シリアルポートの列挙およびNICへの切替を行う。
 */

'use strict';

const log_error = require('debug')('mw-usb-initiator:dpt-serial-port-controller:error');
const log_debug = require('debug')('mw-usb-initiator:dpt-serial-port-controller:debug');
const log_inv = require('debug')('mw-usb-initiator:dpt-serial-port-controller:investigation');

const DPT = require('./dpt-constants');
const DptSerialPort = require('./dpt-serial-port');
const PnpIdHelper = require('./win-pnpid-helper');
const SerialPort = require('serialport-mw');
const retry = require('retry');

/**
 * Win/MacにUSB接続されたDPTの仮想シリアルポート(CDC-ACM)を取り扱うクラス。
 */
class DptSerialPortController {

  /**
   * @constructor
   */
  constructor() {
    this.isRetrying_ = false;
    this.isRetryStopped_ = false;
    this.retryStopCallback = null;
    this.retryRequested_ = false;
  }

  /**
   * PCにUSB接続されているDPTの仮想シリアルポートのポート名の一覧を列挙する。
   * 
   * @param {function(error:Error, comPortArray:Array.<string>)} callback
   *  - DPTの仮想シリアルポートの一覧の取得結果の通知先callback。
   */
  static enum(callback) {
    log_debug('enum()');

    if (typeof callback !== 'function') {
      const err = new TypeError('callback !== function [' + typeof callback + ']');
      log_error(err);
      throw err;
    }

    // シリアルポートの一覧を取得する
    SerialPort.list(function onList(error, portArray) {
      if (error) {
        log_error(error);
        callback(error);
        return;
      }
      if (!Array.isArray(portArray)) {
        const err = new TypeError('portArray is not Array [' + typeof portArray + ']');
        log_error(err);
        callback(err);
        return;
      }

      // DPTのVID,PIDを持つ仮想シリアルポート(CDC-ACM)の一覧を作成する 
      let comPortArray = [];
      for (let i = 0; i < portArray.length; i++) {
        const port = portArray[i];
        if (!port) {
          log_error(new Error('[ERROR] illegal port is found.'));
          continue;
        }

        // シリアルポート情報からVID,PIDを抜き出す 
        const usbVidPid = DptSerialPortController.extractUsbVidPid_(port);
        if (!usbVidPid) {
          log_debug('[WARNING] extractUsbVidPid() failed. ' + JSON.stringify(port));
          continue;
        }
        log_debug('[' + port.comName + '] VID: ' + usbVidPid.vid + ' PID: ' + usbVidPid.pid);

        // DPTのVID,PIDを持つシリアルポートか確認する
        if (usbVidPid.vid === DPT.VID_SONY && usbVidPid.pid === DPT.PID_CDC_ACM) {
          comPortArray.push(port.comName);
        }
      }
      callback(null, comPortArray);
    });
  }

  /**
   * シリアルポートの情報からVendor ID (VID) および Product ID (PID) を取得する。 
   * 
   * @param  {Object} portInfo  - SerialPort#list(callback)で返却されるシリアルポートの情報。
   * @return {Object} usbVidPid - Vendor ID および Product ID。エラー時にはnull。
   * @private
   */
  static extractUsbVidPid_(portInfo) {
    if (!portInfo) {
      log_error(new Error('[ERROR] illegal portInfo is specified.'));
      return null;
    }

    switch (process.platform) {
      case 'win32':
        return PnpIdHelper.extractUsbVidPid(portInfo.pnpId);
      case 'darwin':
        return {
          vid: parseInt(portInfo.vendorId, 16),
          pid: parseInt(portInfo.productId, 16),
        };
      default:
        log_error(new Error('[ERROR] unsupported platform : ' + process.platform));
        return null;
    }
  }

  /**
   * Win/MacにUSB接続されている全てのDPTのシリアルポートをRNDIS/CDC-ECMに切り替える。
   * 
   * @param {function(?Error)} callback
   *    USB切り替えの結果通知用のcallback。
   */
  startSwitchingUsbMode(usbDetector, callback) {
    log_debug('startSwitchingUsbMode()');

    if (typeof callback !== 'function') {
      const err = new TypeError('callback !== function  [' + typeof callback + ']');
      log_error(err);
      throw err;
    }

    const self = this;

    if (self.isRetrying_) {
      self.retryRequested_ = true;
      log_inv('>>> Now Retrying...');
      return;
    }

    self.isRetrying_ = true;
    self.isRetryStopped_ = false;
    self.retryRequested_ = false;

    var done = function () {
      self.isRetrying_ = false;
      self.isRetryStopped_ = true;
      if (self.retryRequested_) {
        self.retryRequested_ = false;
        setTimeout(function () {
          self.startSwitchingUsbMode(usbDetector, callback);
        }, 1000);
      }
      log_inv('>>> Done!');
    };

    var startTime = new Date();

    var tryToSwitchUsbModeOfAllDevices = function () {
      if (self.isRetryStopped_) {
        if (self.retryStopCallback) {
          self.retryStopCallback();
        }
        done();
      } else {
        DptSerialPortController.switchUsbModeOfAllDevices_(usbDetector, function (error) {
          if (!error) {
            callback(null);
            done();
          } else {
            var currentTime = new Date();
            var elapsedTime = currentTime - startTime;
            if (elapsedTime >= (1000 * 60 * 10)) {
              callback(error);
              done();
            } else {
              setTimeout(tryToSwitchUsbModeOfAllDevices, 1000);
            }
          }
        });
      }
    };

    setTimeout(tryToSwitchUsbModeOfAllDevices, 1000);
  }

  /**
   * USB切り替えのリトライ処理を終了させる。
   * 
   * @param {function()} callback
   */
  stopSwithingUsbMode(callback) {
    log_debug('stopSwithingUsbMode()');

    if (typeof callback !== 'function') {
      const err = new TypeError('callback !== function  [' + typeof callback + ']');
      log_error(err);
      throw err;
    }

    const self = this;

    if (!self.isRetrying_) {
      self.isRetryStopped_ = true;
      self.retryStopCallback = null;
      callback();
    } else {
      self.isRetryStopped_ = true;
      self.retryStopCallback = callback;
    }
  }

  /**
   * Win/MacにUSB接続されている全てのDPTのシリアルポートをRNDIS/CDC-ECMに切り替える。
   * 
   * @param {function(error:Error)} callback - USB切り替えの結果通知用のcallback。
   * @private
   */
  static switchUsbModeOfAllDevices_(usbDetector, callback) {
    log_debug('switchUsbModeOfAllDevices_()');

    Promise.resolve()
      .then(function () {
        // ----------------------------------------
        // 1. USB(CDC-ACM)の数をカウント
        // ----------------------------------------
        log_inv('>>> Promise 1');

        return new Promise(function (resolve, reject) {
          log_debug('DptUsbDetector#countUsbSerialPorts()');

          usbDetector.countUsbSerialPorts(function onCount(error, count) {
            if (error) {
              log_error(error);
              reject(error);
            }

            log_debug('DptUsbDetector#countUsbSerialPorts() Count: ' + count);
            resolve(count);
          });
        });
      })
      .then(function (count) {
        // ----------------------------------------
        // 2. シリアルポートの列挙
        // ----------------------------------------
        log_inv('>>> Promise 2');

        return new Promise(function (resolve, reject) {
          // DPTのシリアルポートの一覧を列挙する
          log_debug('DptSerialPortController#enum()');

          DptSerialPortController.enum(function onList(error, portPathArray) {
            if (error) {
              log_error(error);
              reject(error);
              return;
            }

            if (!Array.isArray(portPathArray)) {
              const err = new TypeError('portPathArray is not Array : ' + portPathArray);
              log_error(err);
              reject(err);
              return;
            }

            log_inv('>>> Promise 2  ---  USB:' + count + ', COM:' + portPathArray.length);

            if (count != portPathArray.length) {
              const err = new Error('The number of USB(CDC-ACM) is illegal. USB:' + count + ' COM:' + portPathArray.length);
              log_error(err);
              reject(err);
              return;
            }

            resolve(portPathArray);
          });
        });
      })
      .then(function (portPathArray) {
        // ----------------------------------------
        // 3. USB切替コマンドの発行
        // ----------------------------------------
        log_inv('>>> Promise 3');

        // シリアルポートが検出されなかった場合
        if ((!portPathArray) || (portPathArray.length === 0)) {
          log_debug('COM port is not found.');
          return Promise.resolve();
        }

        // 各シリアルポートにRNDIS/CDC-ECMへの切り替えコマンドを発行する
        const promiseArray = [];
        for (let i = 0; i < portPathArray.length; i++) {
          const portPath = portPathArray[i];
          const promise = new Promise(function (resolve, reject) {
            log_debug('DptSerialPort#switchToEthernet(' + portPath + ')');
            DptSerialPort.switchToEthernet(portPath, function (error) {
              if (error) {
                log_inv('>>> Promise 3  ---  switchToEthernet() failed.');
                log_error(error);
              }

              resolve();
            });
          });

          promiseArray.push(promise);
        }

        // 全てのシリアルポートに対するUSB切り替えを待機させる
        return Promise.all(promiseArray);
      })   
      .then(function () {
        // ----------------------------------------
        // 4. USB切替後に1秒間待機
        // ----------------------------------------
        log_inv('>>> Promise 4');

        return new Promise(function (resolve, reject) {
          setTimeout(resolve, 1000);
        });
      })
      .then(function onFullfilled() {
        // ----------------------------------------
        // 5. USB(CDC-ACM)の数をカウント
        // ----------------------------------------
        log_inv('>>> Promise 5');

        return new Promise(function (resolve, reject) {
          usbDetector.countUsbSerialPorts(function onCount(error, count) {
            if (error) {
              log_error(error);
              reject(error);
              return;
            }

            log_inv('>>> Promise 5  ---  COM:' + count);

            if (count > 0) {
              // シリアルポートが検出された場合
              const err = new Error('USB(CDC-ACM) is found : ' + count);
              log_error(err);
              reject(err);
              return;
            }

            log_debug('DptUsbDetector#countUsbSerialPorts() Count: ' + count);
            resolve(count);
          });
        });
      })  
      .then(function (count) {
        // ----------------------------------------
        // 6. USB切替結果の確認
        // ----------------------------------------
        log_inv('>>> Promise 6');

        return new Promise(function (resolve, reject) {
          // 再度、DPTのシリアルポート一覧を列挙し、切り替わっていることを確認する
          log_debug('DptSerialPortController#enum()');
          DptSerialPortController.enum(function onList(error, portPathArray) {
            if (error) {
              log_error(error);
              reject(error);
              return;
            }

            if (portPathArray && portPathArray.length > 0) {
              // シリアルポートが検出された場合
              const err = new Error('COM port is found. ' + portPathArray);
              log_error(err);
              reject(err);
              return;
            }

            resolve();
          });
        });

      })
      .then(function (value) {
        // ----------------------------------------
        // 7. 結果をcallbackする
        // ----------------------------------------
        log_inv('>>> Promise 7');

        callback(null);
      })
      .catch(function (error) {
        // ----------------------------------------
        // 8. エラー処理
        // ----------------------------------------
        log_inv('>>> Promise 8');

        log_error(error);
        callback(error);
      });
  }
}

module.exports = DptSerialPortController;
